### DISCUSSION TITLE
Can I change .env variables based on NODE_ENV for running production, staging, and development in Nuxt 3?

### DISCUSSION THREAD

**Discussion by Brett-Mulder** [nuxt/nuxt#29653] [Questions]
Hey everyone,

I'm working on a Nuxt 3 project and want to set up different environment variables for production, staging, and development. Here's my nuxt.config.ts:

```typescript
export default defineNuxtConfig({
    devtools: { enabled: false },
    css: ['~/assets/css/main.css'],
    runtimeConfig: {
        authSecret: 'THISisAtestSECRET', // Override it with NUXT_AUTH_SECRET to use an actual key
        public: {
            apiEndpoint: 'http://localhost:4000',
        },
    },
    auth: {
        isEnabled: true,
        baseURL: process.env.NUXT_PUBLIC_AUTH_ORIGIN,
    },
})
```
I would like to have different `.env` files, such as:
- `.env.develop`
- `.env.staging`
- `.env.production`

I want to use `config.public.apiEndpoint` throughout my code, and have it change based on the environment without hardcoding the different endpoints in the `nuxt.config.ts`. For example, using a different API endpoint for production, staging, and development environments.

Is there a way to configure this in Nuxt 3, so it automatically reads from the corresponding `.env` file based on `NODE_ENV`?

If there are other solutions for this, please let me know!

Thanks in advance!

**Comment by xjccc** [comment]
Is this help?
https://nuxt.com/docs/api/commands/dev

**Reply by Brett-Mulder** [reply]
I tried running it with `nuxt start --dotenv .env.production` and (created a new folder because it says path to folder) `nuxt start --dotenv ./config.production/.env`

Using the flag --dotenv does also not work. For testing i added a which:
```typescript
runtimeConfig: {
        authSecret: 'THISisAtestSECRET', // Override it with NUXT_AUTH_SECRET to use an actual key
        public: {
            apiEndpoint: 'http://localhost:4000',
            nodeEnv: process.env.NODE_ENV,
            which: process.env.WHICH,
        },
    },
```

in my .env.production is:
NODE_ENV="production"
WHICH="production"

And my normal .env is:
NODE_ENV="development"
WHICH="development"

And while printing it, the only thing that changed was the `NODE_ENV`. When running `nuxt dev` the `NODE_ENV` is `development`. And when running `nuxt start` the `NODE_ENV` is `production`. But the `WHICH` from the .env file stays always at `development` because that is in the main `.env` file.

**Reply by xjccc** [reply]
https://stackblitz.com/edit/nuxt-starter-zj3jzp?file=package.json,.env,nuxt.config.ts,app.vue

1. `npm run dev` -> `which -> dev`
2. `npm run start` -> `which -> production`

is this right ?

**Reply by Brett-Mulder** [reply]
No, both are `development`. Only `NODE_ENV` changes between `development` and `production`. 

But the strange thing is:
```json
"scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev --host --dotenv .env.production",
    "generate": "nuxt generate",
    "preview": "nuxt preview --dotenv .env.production",
    "postinstall": "nuxt prepare"
  },
```

When running `npm run dev` is picks the .env.production file. But running `npm run preview` is does not pick the `.env.production` file.

**Reply by xjccc** [reply]
There is some explanation.
https://nuxt.com/docs/guide/directory-structure/env#production
